### PR TITLE
Struct mech/porting to atomic utilities

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
Porting to Atomic Utilities

- StructuralMechnicsApplications
- Parallel-OMP

**Changelog**
Replacing loops with omp atomic to use AtomicUtilities

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include "utilities/builtin_timer.h"
 #include "utilities/atomic_utilities.h"
 #include "utilities/entities_utilities.h"
+#include "utilities/atomic_utilities.h"
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/eigensolver_strategy.hpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp [hunk 2]
```diff
@@ -695,12 +694,10 @@ void TrussElement3D2N::AddExplicitContribution(
...
             int index = i * msDimension;
 
-            #pragma omp atomic
-            r_nodal_mass += mass_vector[index];
+            AtomicAdd(r_nodal_mass, mass_vector[index]);
 
             for (int k = 0; k < msDimension; ++k) {
-                #pragma omp atomic
-                r_nodal_inertia[k] += 0.0;
+                AtomicAdd(r_nodal_inertia[k], 0.0);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: @KlausBSautter why did you add this?

File: applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp [hunk 3]
```diff
@@ -1154,12 +1153,10 @@ void MembraneElement::AddExplicitContribution(
...
             SizeType index = i * dimension;
 
-            #pragma omp atomic
-            r_nodal_mass += mass_vector[index];
+            AtomicAdd(r_nodal_mass, mass_vector[index]);
 
             for (SizeType k = 0; k < dimension; ++k) {
-                #pragma omp atomic
-                r_nodal_inertia[k] += 0.0;
+                AtomicAdd(r_nodal_inertia[k], 0.0);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp]: same

File: applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp [hunk 4]
```diff
@@ -1154,12 +1155,10 @@ void MembraneElement::AddExplicitContribution(
...
```
Comment (ashishdarekar) [inline applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp]: What is the use of this line, where we are adding to zero to each element in the array? Or is it just a standard formula where the constant which they are adding is zero here?

File: applications/StructuralMechanicsApplication/custom_processes/shell_to_solid_shell_process.cpp [hunk 5]
```diff
@@ -621,10 +622,12 @@ inline void ShellToSolidShellProcess<TNumNodes>::ComputeNodesMeanNormalModelPart
...
             const array_1d<double, 3>& r_normal = this_geometry.UnitNormal(aux_coords);
             auto& aux_normal = this_node.GetValue(NORMAL);
+
             for (unsigned int index = 0; index < 3; ++index) {
-                #pragma omp atomic
-                aux_normal[index] += r_normal[index];
+                AtomicAdd(aux_normal[index], r_normal[index]);
             }
+            //OR Vector form
+            //AtomicAdd< array_1d<double, 3>, const array_1d<double, 3> >(aux_normal, r_normal);
```
Comment (ashishdarekar) [inline applications/StructuralMechanicsApplication/custom_processes/shell_to_solid_shell_process.cpp]: This is another way (using vector type AtomicAdd()) , will it work ??

Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_processes/shell_to_solid_shell_process.cpp]: It would (I think) but we might change this fct a bit hence I would not use it for now

File: applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp (hunk 4 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp]: @KlausBSautter

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp (hunk 2 above)
Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: because no nodal_inertia is added. While I am typing these words I wonder why we do not just skip this :D

File: applications/StructuralMechanicsApplication/custom_processes/shell_to_solid_shell_process.cpp (hunk 5 above)
Comment (ashishdarekar) [reply applications/StructuralMechanicsApplication/custom_processes/shell_to_solid_shell_process.cpp]: Okay, Fine.

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp (hunk 2 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: so can we remove it then?
I mean it does nothing...

Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: I guess so .....

Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: ok then @ashishdarekar please go ahead and remove it :)

Comment (ashishdarekar) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: Done :)

**Comment by philbucher** [comment]
@ashishdarekar now #8155 is also merged so you can use it here :)

File: applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp [hunk 6]
```diff
@@ -1154,13 +1153,7 @@ void MembraneElement::AddExplicitContribution(
             array_1d<double, 3>& r_nodal_inertia = GetGeometry()[i].GetValue(NODAL_INERTIA);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/membrane_element.cpp]: @KlausBSautter maybe you can take another look, now `r_nodal_inertia` is unused, you sure it is correct?

**Review by KlausBSautter — DISMISSED** [review]
I checked my local tests and it seems to work fine

**Review by philbucher — APPROVED** [review]
